src/server/HeatGenerator.py

```diff
@@ -0,0 +1,131 @@
+                'auto_frequency': True,
+                'defaultMethod': ProgramMethod.NONE
+                })
+            heat_id_mapping.append(new_heat.id)
+
+        for h_idx, heat_plan in enumerate(generated_heats):
+            heat_alterations = []
+            heat_slots = self._RHData.get_heatNodes_by_heat(heat_id_mapping[h_idx])
+
+            if len(heat_slots) < len(heat_plan['slots']):
+                logger.warning('Not enough actual slots for requested heat generation')
+
+            for s_idx, heat_slot in enumerate(heat_slots):
+                if s_idx < len(heat_plan['slots']):
+                    seed_slot = heat_plan['slots'][s_idx]
+                    data = {
+                        'heat': heat_id_mapping[h_idx],
+                        'slot_id': heat_slot.id,
+                        'seed_rank': seed_slot['seed_rank'],
+                        }
+                    if seed_slot['method'] == 'input':
+                        if input_class:
+                            data['method'] = ProgramMethod.CLASS_RESULT
+                            data['seed_class_id'] = input_class
+                        else:
+                            # randomly seed 
+                            if filled_pool == False:
+                                for pilot in self._RHData.get_pilots():
+                                    pilot_pool.append(pilot.id)
+
+                                random.shuffle(pilot_pool)
+                                filled_pool = True
+
+                            if len(pilot_pool):
+                                data['method'] = ProgramMethod.ASSIGN
+                                data['pilot'] = pilot_pool.pop()
+                            else:
+                                logger.info('Unable to seed pilot: no available pilots left to seed')
+                                data['method'] = ProgramMethod.NONE
+
+                    elif seed_slot['method'] == ProgramMethod.HEAT_RESULT:
+                        data['method'] = ProgramMethod.HEAT_RESULT
+                        data['seed_heat_id'] = heat_id_mapping[seed_slot['seed_heat_id']]
+                    else:
+                        logger.error('Not a supported seed method: {}'.format(seed_slot['method']))
+                        return False
+
+                heat_alterations.append(data)
+
+            self._RHData.alter_heatNodes_fast(heat_alterations)
+
+        if filled_pool and len(pilot_pool):
+            logger.info("{} unseeded pilots remaining in pool".format(len(pilot_pool)))
+
+class HeatGenerator():
+    def __init__(self, name, label, generatorFn, defaultArgs=None):
+        self.name = name
+        self.label = label
+        self.generator = generatorFn
+        self.defaultArgs = defaultArgs
+
+    def generate(self, RHData, Results, PageCache, generate_args=None):
+        if generate_args and self.defaultArgs:
+            generate_args = {**self.defaultArgs, **generate_args}
```

Comment:
This looks a bit like a potential bug: if there are default argumenst, why use them only conditionally on generate_args not being False? A simple

```
generate_args = {**self.defaultArgs, **generate_args if generate_args else {}}
```
should do it, and not create a situation where the default args are ignored in case no explicit args are used.